### PR TITLE
Change `ed19` to work with arbitrary-length messages

### PR DESCRIPTION
**Pull request by xgreenx**
Closes https://github.com/FuelLabs/fuel-vm/issues/793. Spec PR https://github.com/FuelLabs/fuel-specs/pull/600

## Checklist
- [x] Breaking changes are clearly marked as such in the PR description and changelog

### Before requesting review
- [x] I have reviewed the code myself

### CODE REVIEW

File: fuel-vm/src/tests/crypto.rs [hunk 1]
```diff
@@ -601,7 +616,7 @@ fn ed25519_verify_c_gt_vmaxram_sub_32() {
         op::xor(reg_b, reg_b, reg_b),
         op::not(reg_a, RegId::ZERO),
         op::subi(reg_a, reg_a, 31),
-        op::ed19(reg_b, reg_b, reg_a),
+        op::ed19(reg_b, reg_b, reg_a, 32),
```
Comment (mchristopher) [inline fuel-vm/src/tests/crypto.rs:619]: Can we add 2 more tests, one to check 64-byte len, and one where we leave the len blank/empty?

Comment (Dentosal) [reply fuel-vm/src/tests/crypto.rs]: Added in f964db61

File: fuel-vm/src/tests/crypto.rs [hunk 2]
```diff
@@ -578,30 +650,38 @@ fn ed25519_verify_a_gt_vmaxram_sub_64() {
...
 }
 
-#[test]
-fn ed25519_verify_c_gt_vmaxram_sub_32() {
+#[test_case(31, 32 => (); "Just over the end with 32 bits")]
+#[test_case(63, 64 => (); "Just over the end with 64 bits")]
+#[test_case(31, 0 => (); "Zero defaults to 32")]
+#[test_case(31, 100 => (); "Way over the end")]
+#[test_case(0, 32 => (); "Empty range, goes over it")]
+fn ed25519_verify_c_plus_d_gt_vmaxram(offset: u16, len: u32) {
```
Comment (MitchTurner) [inline fuel-vm/src/tests/crypto.rs:673]: I feel like the name should mention that we should expect it to "overflow". All of these tests have to do with the offset causing the overflow, so maybe the name should reflect that.

The `test_case` names don't really inform me that much either, but I think that has more to do with the base name not being clear enough.

File: fuel-tx/src/transaction/consensus_parameters/gas.rs [hunk 3]
```diff
@@ -945,6 +936,24 @@ impl GasCostsValues {
...
+            },
+            GasCostsValues::V2(v2) => DependentCost::HeavyOperation {
+                base: v2.ed19,
+                gas_per_unit: 0,
+            },
+            GasCostsValues::V3(v3) => DependentCost::HeavyOperation {
+                base: v3.ed19,
+                gas_per_unit: 0,
+            },
+            GasCostsValues::V4(v4) => v4.ed19,
```
[the remark this answers is not in the document]
Comment (Dentosal) [reply fuel-tx/src/transaction/consensus_parameters/gas.rs:953]: Probably just an oversight. Brandon was wonderin which one it should be, and changing it seemed to make sense. https://github.com/FuelLabs/fuel-vm/pull/795#discussion_r1690102472

File: fuel-vm/src/interpreter/executors/instruction.rs [hunk 4]
```diff
@@ -59,7 +59,7 @@ where
                 ))
             })?,
         );
-        if pc < self.registers[RegId::IS] || pc >= self.registers[RegId::SSP] {
+        if pc < self.registers[RegId::IS] || pc >= self.registers[RegId::SP] {
```
[the remark this answers is not in the document]
Comment (Dentosal) [reply fuel-vm/src/interpreter/executors/instruction.rs]: The PR originially included changes from an old version of the Blob TX PR. Git merged this part incorrectly and I missed it when reviewing the code myself. Removed in 1c56802e.

File: fuel-vm/src/tests/crypto.rs (hunk 2 above)
Comment (Dentosal) [reply fuel-vm/src/tests/crypto.rs]: The other tests were named like this as well, I was trying to be consistent. But I just fixed them all in eb118178 now.

File: fuel-vm/src/interpreter/executors/instruction.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Dentosal) [reply fuel-vm/src/interpreter/executors/instruction.rs]: I re-reviewed myself as well to make sure nothing else was missed.

**Review by MitchTurner — APPROVED** [review]
Looks great. Thanks for the changes, @Dentosal
